src/platform_specific/affinity.unix.cc

```diff
@@ -1,20 +1,206 @@
+// Calls to pin_this_thread prior to this call will have no effect
+void initialize(const runtime_configuration& cfg) {
+	std::lock_guard lock(g_state.mutex);
+	assert(!g_state.initialized && "Thread pinning already initialized.");
+	assert(g_state.thread_pinning_plan.empty() && "Thread pinning plan not initially empty.");
+	assert(g_state.pinned_threads.empty() && "Pinned threads not initially empty.");
+
+	g_state.config = cfg;
+
+	const auto ret = sched_getaffinity(0, sizeof(cpu_set_t), &g_state.available_cores);
+	if(ret != 0) {
+		CELERITY_WARN("Error retrieving initial process affinity mask, will disable pinning.");
+		g_state.config.enabled = false;
+		return;
 	}
 
-} // namespace detail
-} // namespace celerity
+	// pinned threads per process: user, scheduler, executor, 1 backend worker per device
+	uint32_t pinned_threads_per_process = 3 + g_state.config.num_devices;
+	// total number of threads to be pinned across processes (legacy mode)
+	uint32_t total_threads = pinned_threads_per_process * g_state.config.num_legacy_processes;
+
+	if(g_state.config.enabled) {
+		// build our pinning plan
+		if(!cfg.hardcoded_core_ids.empty()) {
+			// just use the provided hardcoded IDs
+			if(static_cast<uint32_t>(cfg.hardcoded_core_ids.size()) != total_threads) {
+				CELERITY_WARN("Hardcoded core ID count ({}) does not match the number of threads to be pinned ({}), will disable pinning.",
+				    cfg.hardcoded_core_ids.size(), total_threads);
+				g_state.config.enabled = false;
+				return;
+			}
+
+			uint32_t core_id = cfg.legacy_process_index * pinned_threads_per_process;
+			g_state.thread_pinning_plan.emplace(thread_type::user, cfg.hardcoded_core_ids[core_id++]);
+			g_state.thread_pinning_plan.emplace(thread_type::scheduler, cfg.hardcoded_core_ids[core_id++]);
+			g_state.thread_pinning_plan.emplace(thread_type::executor, cfg.hardcoded_core_ids[core_id++]);
+			for(uint32_t i = 0; i < g_state.config.num_devices; ++i) {
+				g_state.thread_pinning_plan.emplace(static_cast<thread_type>(thread_type::first_backend_worker + i), cfg.hardcoded_core_ids[core_id++]);
+			}
+		} else {
+			// otherwise, sequential core assignments for now; it is most important that each of the threads is "close"
+			// to the ones next to it in this sequence, so that communication between them is fast
+
+			uint32_t current_core = g_state.config.standard_core_start_id;
+			uint32_t assigned_cores = 0;
+
+			// find the next sequential core we may use
+			auto next_available_core = [&]() {
+				if(!g_state.config.enabled) return 0u;
+				while(CPU_ISSET(current_core, &g_state.available_cores) == 0 && current_core < CPU_SETSIZE) {
+					current_core++;
+				}
+				if(current_core >= CPU_SETSIZE) {
+					CELERITY_WARN("Ran out of available cores for thread pinning after assigning {} cores, will disable pinning.", assigned_cores);
+					g_state.config.enabled = false;
+					return 0u;
```

Comment:
This is oddly non-local, maybe `throw std::out_of_range` here and surround L105-L115 with a try-catch? Would also get rid of warning spam.

Reply:
Refactored this whole thing to be much better, so this no longer exists.